### PR TITLE
[23.x] Backport some changes necessary for google-cloud-cpp resp. conda-forge

### PR DESCRIPTION
**Pull request by h-vetinari**
I don't know what the procedure is on backporting to maintenance branches, but I thought I'd try by opening a PR.

In conda-forge, we'd definitely need:
* https://github.com/protocolbuffers/protobuf/pull/12903 by @coryan (because google-cloud-cpp will otherwise [not compile](https://github.com/googleapis/google-cloud-cpp/issues/8125) with the 23.x series)
* https://github.com/protocolbuffers/protobuf/pull/12978 by @traversaro (because now protobuf requires abseil and all consumers need to link to it + the fact that we build everything based on shared abseil libraries, including on windows)
* we also need https://github.com/protocolbuffers/protobuf/commit/fc1c5512e524e0c00a276aa9a38b2cdb8fdf45c7 for https://github.com/protocolbuffers/protobuf/issues/12932

CC @fowles @mkruskal-google

### CODE REVIEW

File: src/google/protobuf/port_def.inc [hunk 1]
```diff
@@ -888,11 +888,18 @@ static_assert(PROTOBUF_ABSL_MIN(20230125, 3),
...
-// autoheader defines this in some circumstances
-#ifdef PACKAGE
-#define PROTOBUF_DID_UNDEF_PACKAGE
-#pragma push_macro("PACKAGE")
-#undef PACKAGE
+// a few common headers define this
+#ifdef PACKED
+#define PROTOBUF_DID_UNDEF_PACKED
+#pragma push_macro("PACKED")
+#undef PACKED
```
Comment (h-vetinari) [inline src/google/protobuf/port_def.inc:895]: It's worth noting that the section for dealing with `PACKAGE` is duplicated and that duplication got removed in #12903.

`PACKED` got added by https://github.com/protocolbuffers/protobuf/commit/7dbe742e6c4bea916c54713e31811412d2076351, which is effectively backported here through the conflict resolution.

File: src/google/protobuf/port_undef.inc [hunk 2]
```diff
@@ -128,14 +128,23 @@
 #undef PROTOBUF_DID_UNDEF_PACKAGE
 #endif
 
+#ifdef PROTOBUF_DID_UNDEF_PACKED
+#pragma pop_macro("PACKED")
+#undef PROTOBUF_DID_UNDEF_PACKED
+#endif
```
Comment (h-vetinari) [inline src/google/protobuf/port_undef.inc:134]: Same here w.r.t. https://github.com/protocolbuffers/protobuf/commit/7dbe742e6c4bea916c54713e31811412d2076351

**Review by h-vetinari — COMMENTED** [review]
There were some trivial conflicts here that I resolved while backporting #12903. Here's some short comments about where they appeared.

File: src/google/protobuf/port_def.inc (hunk 1 above)
Comment (fowles) [reply src/google/protobuf/port_def.inc:895]: Thanks

**Review by fowles — APPROVED** [review]

**Comment by h-vetinari** [comment]
Forgot one: we also need https://github.com/protocolbuffers/protobuf/commit/fc1c5512e524e0c00a276aa9a38b2cdb8fdf45c7 for https://github.com/protocolbuffers/protobuf/issues/12932

**Review by fowles — APPROVED** [review]

**Pull request merged**